src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,686 @@
+                      len, ret, errno);
+            status = UCS_ERR_IO_ERROR;
+            goto terminate;
+        }
+        /* EINVAL: no length info, ENOSPC: needed length is set */
+        if (errno == EINVAL) {
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            ref_len = len;
+        }
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (!umad) {
+            return UCS_ERR_NO_MEMORY;
+        }
+        goto retry;
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto terminate;
+    }
+
+    len -= IB_VENDOR_RANGE2_DATA_OFFS;
+    if (len <= 0) {
+        if (user_mad->status == ETIMEDOUT) {
+            ucs_info("MAD: receive: Remote unreachable");
+            status = UCS_ERR_UNREACHABLE;
+        } else {
+            status = UCS_ERR_OUT_OF_RANGE;
+        }
+        goto terminate;
+    }
+
+    ret = umad_status(umad);
+    if (ret) {
+        ucs_info("MAD: receive: Status failure: %d", ret);
+        status = UCS_ERR_REJECTED;
+        goto terminate;
+    }
+    status = UCS_OK;
+
+    data = (uint8_t*)umad_get_mad(umad) + IB_VENDOR_RANGE2_DATA_OFFS;
+    if (perftest_mad_is_ping(data, len)) {
+        ucs_debug("PONG");
+        goto retry;
+    }
+
+    if (len > *avail) {
+        status = UCS_ERR_MESSAGE_TRUNCATED;
+        len    = *avail;
+    }
+    memcpy(buffer, data, len);
+    *avail = len;
+terminate:
+    free(umad);
+    return status;
+}
+
+static ucs_status_t
+perftest_mad_recv_from_remote(perftest_mad_rte_group_t *rte_group, void *buffer,
+                              size_t *avail, const ib_portid_t *target_port)
+{
+    ucs_status_t ret        = UCS_ERR_IO_ERROR;
+    ib_portid_t remote_port = {
+        .lid = 0
+    };
+    size_t size;
+
+    while ((remote_port.lid == 0) || (remote_port.lid != target_port->lid)) {
+        size            = *avail;
+        remote_port.lid = 0;
+
+        ret = perftest_mad_recv(rte_group, buffer, &size, &remote_port);
+        if (ret == UCS_ERR_NO_PROGRESS) {
+            ucs_debug("PING");
+            ret = perftest_mad_ping(rte_group);
+            ucs_assertv(ret == UCS_OK, "ret: %s", ucs_status_string(ret));
+            remote_port.lid = 0;
+            continue;
+        }
+        if (ret == UCS_ERR_UNREACHABLE) {
+            ucs_error("MAD: recv: Remote unreachable");
+            exit(EXIT_FAILURE);
+        }
+    }
+    ucs_info("MAD: recv packet size:%zu/%zu", size, *avail);
+    *avail = size;
+    return ret;
+}
+
+static unsigned rte_mad_group_size(void *rte_group)
+{
+    return 2;
+}
+
+static unsigned rte_mad_group_index(void *rte_group)
+{
+    return !((perftest_mad_rte_group_t*)rte_group)->is_server;
+}
+
+static ucs_status_t
+perftest_mad_recv_magic(perftest_mad_rte_group_t *group, unsigned value)
+{
+    unsigned snc = 0;
+    size_t size  = sizeof(snc);
```

Comment:
is initialization needed because of coverity error? If yes, looks like better to initialize it with 0

Reply:
snc is received from the peer so we have set it to expected receive size.